lib/internal/encoding.js

```diff
@@ -0,0 +1,359 @@
+  ['x-mac-ukrainian', 'x-mac-cyrillic'],
+  ['chinese', 'gbk'],
+  ['csgb2312', 'gbk'],
+  ['csiso58gb231280', 'gbk'],
+  ['gb2312', 'gbk'],
+  ['gb_2312', 'gbk'],
+  ['gb_2312-80', 'gbk'],
+  ['gbk', 'gbk'],
+  ['iso-ir-58', 'gbk'],
+  ['x-gbk', 'gbk'],
+  ['gb18030', 'gb18030'],
+  ['big5', 'big5'],
+  ['big5-hkscs', 'big5'],
+  ['cn-big5', 'big5'],
+  ['csbig5', 'big5'],
+  ['x-x-big5', 'big5'],
+  ['cseucpkdfmtjapanese', 'euc-jp'],
+  ['euc-jp', 'euc-jp'],
+  ['x-euc-jp', 'euc-jp'],
+  ['csiso2022jp', 'iso-2022-jp'],
+  ['iso-2022-jp', 'iso-2022-jp'],
+  ['csshiftjis', 'shift_jis'],
+  ['ms932', 'shift_jis'],
+  ['ms_kanji', 'shift_jis'],
+  ['shift-jis', 'shift_jis'],
+  ['shift_jis', 'shift_jis'],
+  ['sjis', 'shift_jis'],
+  ['windows-31j', 'shift_jis'],
+  ['x-sjis', 'shift_jis'],
+  ['cseuckr', 'euc-kr'],
+  ['csksc56011987', 'euc-kr'],
+  ['euc-kr', 'euc-kr'],
+  ['iso-ir-149', 'euc-kr'],
+  ['korean', 'euc-kr'],
+  ['ks_c_5601-1987', 'euc-kr'],
+  ['ks_c_5601-1989', 'euc-kr'],
+  ['ksc5601', 'euc-kr'],
+  ['ksc_5601', 'euc-kr'],
+  ['windows-949', 'euc-kr'],
+  ['utf-16be', 'utf-16be'],
+  ['utf-16le', 'utf-16le'],
+  ['utf-16', 'utf-16le'],
+  // Unsupported standard mappings
+  // ['csiso2022kr', 'replacement'],
+  // ['hz-gb-2312', 'replacement'],
+  // ['iso-2022-cn', 'replacement'],
+  // ['iso-2022-cn-ext', 'replacement'],
+  // ['iso-2022-kr', 'replacement'],
+  // ['x-user-defined', 'x-user-defined'],
+  // ['iso-8859-16', 'iso-8859-16'],
+]);
+
+function normalizeEncoding(encoding) {
+  if (encoding === undefined)
+    return 'utf-8';
+  let enc = encodings.get(encoding);
+  if (enc !== undefined) return enc;
+  enc = encodings.get(encoding.toLowerCase());
+  if (enc !== undefined) return enc;
+}
+
+function hasTextDecoder(encoding = 'utf-8') {
+  if (typeof encoding !== 'string')
+    throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+  return hasConverter(normalizeEncoding(encoding));
+}
+
+class TextDecoder {
+  constructor(encoding = 'utf-8', options = {}) {
+    if (typeof encoding !== 'string')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+    if (options !== undefined && typeof options !== 'object')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'options', 'object');
+
+    const fatal = options !== null &&
+                  options.fatal === true;
+
+    const ignoreBOM = options !== null &&
+                      options.ignoreBOM === true;
+
+    const enc = normalizeEncoding(encoding);
+    if (enc === undefined || enc === 'replacement')
+      throw new errors.RangeError('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    var flags = 0;
+    flags |= fatal ? CONVERTER_FLAGS_FATAL : 0;
+    flags |= ignoreBOM ? CONVERTER_FLAGS_IGNORE_BOM : 0;
+
+    const handle = getConverter(enc, flags);
+    if (handle === undefined)
+      throw new errors.Error('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    Object.defineProperties(this, {
```

Comment:
`Object.defineProperties` is pretty slow to be used on per object construction. I don't see any harm in making `kHandle` configurable, enumerable, or writable, as it's a symbol property already (and it's what we do for URL). `encoding`, `fatal`, and `ignoreBOM` should all be defined on the prototype as getter functions per spec.